Automattic/wp-calypso
# composite-checkout: Include extra domain fields (4)

This has been broken into several smaller PRs; the final one is #41819.

To do:
- [x] Figure out why validation is hanging
- [x] Make sure validation catches errors on the extra fields
- [x] Merge #40575
- [x] Merge #41030
- [x] Merge #40722
- [x] Merge #41033
- [x] Merge #41053
- [x] Merge #41055
- [x] Merge #41058
- [x] Merge #41103
- [x] Merge #41158
- [x] Add missing CA validator to the backend requiring nonempty organization for corporations - D42211-code
- [x] Add missing FR validator to the backend requiring nonempty organization for corporations - D41998-code
- [x] Merge D42658-code
- [ ] Merge #41813
- [ ] Check that the special domain flows work
- [ ] Check that the plan-only flow and non-special domain flows are not broken
- [x] Clean up the commit history

#### Changes proposed in this Pull Request

* Update composite checkout's data model to accommodate the extra contact details required by specific TLDs, and inject calypso's UI components for manipulating these. Depends on #40722.

#### Testing instructions

This pr is not ready for prime time. But when it is, you can test it like this:

Run local calypso with `ENABLE_FEATURES=composite-checkout-force npm start` to force composite checkout.

* Enter checkout with one of the following TLDs in your cart: (1) .ca, (2) .uk, (3) .fr. At the contact details step you should see extra form fields depending on which TLD you have.
* Enter checkout with two or more of the above listed TLDs. You should see extra fields for each TLD.
* Enter checkout with a domain not belonging to one of the special TLDs; make sure you see only the usual domain contact form.
* Enter checkout without a domain; make sure you see only the usual non-domain contact form.

For all test cases proceed at least to the last step to check that validation is successful. Not yet sure how to test completing a purchase with a non-.live domain.
